File: src/gui/backends/sdl2/window.rs
use sdl2::render::{Canvas, TextureCreator, Texture};
use sdl2::video::WindowContext;

use crate::gui::window_flags::WindowFlags;
use crate::gui::color::{self, Color};

use super::{Application, Button};
use super::widget::WidgetType;

pub struct Window<'a>
{
    canvas: Canvas<sdl2::video::Window>,
    texture_creator: TextureCreator<WindowContext>,
    background_color: Color,
    widgets: Vec<(WidgetType<'a>, Vec<&'a Texture<'a>>)>,
}

impl<'a> PartialEq for Window<'a>
{
    fn eq(&self, other: &Self) -> bool {
        if self.canvas.info() == other.canvas.info()
        {
            return true
        }
        else
        {
            return false
        }
    }
}

impl<'a> Window<'a>
{
    pub fn new(app: &Application, title: &str, width: u32, height: u32, background_color: color::Color, flags: WindowFlags) -> Window<'a>
    {
        let mut window_builder = app.video_subsystem.window(title, width, height);

        Self::init_window_flags(flags, &mut window_builder);

        let window = window_builder.build().unwrap();

        let mut canvas = window
            .into_canvas()
            .accelerated()
            .present_vsync()
            .build()
            .unwrap();
        canvas.set_draw_color(background_color.rgba());

        let texture_creator = canvas.texture_creator();

        Window { canvas, texture_creator, background_color, widgets: vec![] }
    }

    pub fn set_fullscreen(&mut self, fullscreen: bool)
    {
        self.canvas.window_mut().set_fullscreen(
            if fullscreen { sdl2::video::FullscreenType::True } else { sdl2::video::FullscreenType::Off }
        )
        .map_err(|e| eprintln!("Error: {e}"));
    }

    pub fn set_bordered(&mut self, bordered: bool)
    {
        self.canvas.window_mut().set_bordered(bordered);
    }

    pub fn set_position(&mut self, x: u32, y: u32)
    {
        use sdl2::video::WindowPos;
        self.canvas.window_mut().set_position(
            WindowPos::Positioned(x as i32),
            WindowPos::Positioned(y as i32)
        );
    }

    pub fn set_size(&mut self, width: u32, height: u32)
    {
        self.canvas.window_mut().set_size(width, height)
        .map_err(|e| eprintln!("Error: {e}"));
    }

    pub fn set_opacity(&mut self, opacity: f32)
    {
        self.canvas
            .window_mut()
            .set_opacity(opacity)
            .map_err(|e| eprintln!("Error: {e}"));
    }

    pub fn set_title(&mut self, new_title: &str)
    {
        self.canvas
            .window_mut()
            .set_title(new_title)
            .map_err(|e| eprintln!("Error: {e}"));
    }

    pub fn add_widget(&'a mut self, widget_type: WidgetType<'a>)
    {
        let mut textures: Vec<&Texture> = vec![];
        for surface in &widget_type.raw().surfaces
        {
            textures.push(&surface.as_texture(&self.texture_creator).unwrap());
        }

        self.widgets.push((widget_type, textures));
    }

    pub fn add_button(&'a mut self, button: &'a Button)
    {
        let mut textures: Vec<&Texture> = vec![];
        for surface in &button.raw().surfaces
        {
            textures.push(&surface.as_texture(&self.texture_creator).unwrap());
        }

        self.widgets.push((WidgetType::Button(button), textures));
    }

    pub fn draw(&mut self)
    {
        self.canvas.set_draw_color(self.background_color.rgb());
        self.canvas.clear();
        // others will be here
        self.widgets.iter().for_each(|widget|{
            let widget_raw = widget.0.raw();
            let textures = &widget.1;

            widget_raw.raw.iter().for_each(|rect| {
                self.canvas.set_draw_color(rect.color.rgba());
                let rect: sdl2::rect::Rect = rect.into();
                self.canvas.fill_rect(rect).map_err(|e| eprintln!("Error: {e}"));
            });

            textures.iter().for_each(|texture| {
                // texture.
            });
        });
        self.canvas.present();
    }
}

impl<'a> Window<'a>
{
    fn init_window_flags(flags: WindowFlags, window_builder: &mut sdl2::video::WindowBuilder)
    {
        if flags & WindowFlags::Fullscreen != WindowFlags::None {
            window_builder.fullscreen();
        };

        if flags & WindowFlags::Shown != WindowFlags::None {
            // it's shown by default
        };

        if flags & WindowFlags::Hidden != WindowFlags::None {
            window_builder.hidden();
        };

        if flags & WindowFlags::Borderless != WindowFlags::None {
            window_builder.borderless();
        };

        if flags & WindowFlags::Resizable != WindowFlags::None {
            window_builder.resizable();
        };

        if flags & WindowFlags::Minimized != WindowFlags::None {
            window_builder.minimized();
        };

        if flags & WindowFlags::Maximized != WindowFlags::None {
            window_builder.maximized();
        };
    }
}

#[cfg(test)]
mod test
{
    // use crate::gui::{window::WindowFlags, color};

    use super::Window;
    use super::Application;

    #[test]
    fn test_run()
    {
        // let mut app = Application::new().unwrap();
        // let mut window = Window::new(&app, "rust-sdl2 demo", 800, 600, color::Color::WHITE, WindowFlags::Default);

        // app.draw_window(&mut window);

        // app.main_loop()
    }
}

File: src/gui/backends/sdl2/window.zig
const sdl2 = @import("sdl2.zig");

const std = @import("std");

const Color = @import("../../color.zig");
const WindowFlags = @import("../../window_flags.zig").WindowFlags;
const widget_mod = @import("widget.zig");
const Button = @import("button.zig");

pub const WindowError = error{
    WINDOW_CREATION_FAILED,
    RENDERER_CREATION_FAILED,
};

const WindowWidget = struct 
{
    widget: widget_mod.WidgetType,
    textures: std.ArrayList(widget_mod.WidgetTexture)
};

pub const Window = struct {
    const Self = @This();

    raw: *sdl2.SDL_Window,
    renderer: *sdl2.SDL_Renderer,
    background_color: Color,
    widgets: std.ArrayList(WindowWidget),
    allocator: *const std.mem.Allocator,

    pub fn new(title: []const u8, width: u32, height: u32, background_color: Color, flags: WindowFlags, allocator: *const std.mem.Allocator) WindowError!Window {

        // window
        var win: ?*sdl2.SDL_Window = sdl2.SDL_CreateWindow(@ptrCast([*c]const u8, title), @intCast(c_int, 100), @intCast(c_int, 100), @intCast(c_int, width), @intCast(c_int, height), flags.as_u32());

        if (win == null) {
            std.log.err("SDL_CreateWindow Error: {s}\n", .{sdl2.SDL_GetError()});
            return WindowError.WINDOW_CREATION_FAILED;
        }

        // render
        var ren: ?*sdl2.SDL_Renderer = sdl2.SDL_CreateRenderer(win, -1, sdl2.SDL_RENDERER_ACCELERATED | sdl2.SDL_RENDERER_PRESENTVSYNC);
        if (ren == null) {
            std.log.err("SDL_CreateRenderer Error: {s}\n", .{sdl2.SDL_GetError()});
            return WindowError.RENDERER_CREATION_FAILED;
        }

        return Window {
            .raw = win.?,
            .renderer = ren.?,
            .background_color = background_color,
            .widgets = std.ArrayList(WindowWidget).init(allocator.*),
            .allocator = allocator,
        };
    }

    pub fn add_widget(self: *Self, widget_type: *widget_mod.WidgetType) !void
    {
        var textures = widget_type.raw().to_textures(self, self.allocator);
        self.widgets.widgets.append(widget_type) catch unreachable;
        self.widgets.textures.?.append(textures) catch unreachable;
    }

    pub fn add_button(self: *Self, button: *Button) void
    {
        var widget_type = widget_mod.WidgetType { .Button = button };
        var textures = widget_type.raw().to_textures(self, self.allocator.*) catch unreachable;
        self.widgets.append( WindowWidget { .widget = widget_type, .textures = textures }) catch unreachable;
    }

    pub fn draw(self: Self) void {
        _ = sdl2.SDL_SetRenderDrawColor(self.renderer, self.background_color.r, self.background_color.g, self.background_color.b, self.background_color.a);
        _ = sdl2.SDL_RenderClear(self.renderer);

        // others go here
        for (self.widgets.items) |*window_widget| {
            switch (window_widget.widget) {
                inline else => |widget| {
                    for (widget.raw().raw_rects.items) |rectangle| {
                        _ = sdl2.SDL_SetRenderDrawColor(
                            self.renderer,
                            rectangle.color.r,
                            rectangle.color.g,
                            rectangle.color.b,
                            rectangle.color.a
                        );

                        var rect = sdl2.SDL_Rect { 
                            .x = @intCast(c_int, rectangle.position.x),
                            .y = @intCast(c_int, rectangle.position.y),
                            .w = @intCast(c_int, rectangle.size.width),
                            .h = @intCast(c_int, rectangle.size.height),
                        };

                        _ = sdl2.SDL_RenderFillRect(self.renderer, &rect);

                        for (window_widget.textures.items) |texture|
                        {
                            _ = sdl2.SDL_RenderCopy(self.renderer, texture.raw, null, &texture.rect);
                        }
                    }
                },
            }
        }

        sdl2.SDL_RenderPresent(self.renderer);
    }

    pub fn destroy(self: Self) void {
        sdl2.SDL_DestroyRenderer(self.renderer);
        sdl2.SDL_DestroyWindow(self.raw);

        for (self.widgets.items) |widget| {
            widget.textures.deinit();
        }
        self.widgets.deinit();
    }
};

test "test window" {
    const allocator = std.testing.allocator;

    var win = try Window.new("This is a test", 800, 600, Color.YELLOW, WindowFlags { .Shown = true }, &allocator);
    win.destroy();
}
